pkg/p2p/server.go

```diff
@@ -0,0 +1,845 @@
+	handlers map[Topic]workerpool.EventHandle
+
+	peerLock sync.RWMutex
+	peers    map[string]*cdcPeer // all currently connected clients
+
+	// messages for topics with NO registered handle.
+	// This can happen when the server is slow.
+	// The upper limit of pending messages is restricted by
+	// MaxTopicPendingCount in MessageServerConfig.
+	pendingMessages map[topicSenderPair][]pendingMessageEntry
+
+	// We store the last processed message's Sequence so that
+	// we can periodically send ACK messages.
+	acksMapLock sync.RWMutex
+	acksMap     map[NodeID]map[Topic]Seq
+
+	// taskQueue is used to store internal tasks MessageServer
+	// needs to execute serially.
+	taskQueue chan interface{}
+
+	// The WorkerPool instance used to execute message handlers.
+	pool workerpool.WorkerPool
+
+	isRunning int32 // atomic
+	closeCh   chan struct{}
+
+	config *MessageServerConfig // read only
+}
+
+type taskOnMessageBatch struct {
+	streamMeta     *p2p.StreamMeta
+	messageEntries []*p2p.MessageEntry
+}
+
+type taskOnRegisterPeer struct {
+	sender     *streamHandle
+	clientAddr string // for logging
+}
+
+type taskOnRegisterHandler struct {
+	topic   string
+	handler workerpool.EventHandle
+	done    chan struct{}
+}
+
+type taskOnDeregisterHandler struct {
+	topic string
+	done  chan struct{}
+}
+
+// taskDebugDelay is used in unit tests to artificially block the main
+// goroutine of the server. It is not used in other places.
+type taskDebugDelay struct {
+	doneCh chan struct{}
+}
+
+// NewMessageServer creates a new MessageServer
+func NewMessageServer(serverID NodeID, config *MessageServerConfig) *MessageServer {
+	return &MessageServer{
+		serverID:        serverID,
+		handlers:        make(map[string]workerpool.EventHandle),
+		peers:           make(map[string]*cdcPeer),
+		pendingMessages: make(map[topicSenderPair][]pendingMessageEntry),
+		acksMap:         make(map[NodeID]map[Topic]Seq),
+		taskQueue:       make(chan interface{}, config.MaxPendingTaskCount),
+		pool:            workerpool.NewDefaultWorkerPool(config.WorkerPoolSize),
+		closeCh:         make(chan struct{}),
+		config:          config,
+	}
+}
+
+// Run starts the MessageServer's worker goroutines.
+// It must be running to provide the gRPC service.
+func (m *MessageServer) Run(ctx context.Context) error {
+	atomic.StoreInt32(&m.isRunning, 1)
+	defer func() {
+		atomic.StoreInt32(&m.isRunning, 0)
+		close(m.closeCh)
+	}()
+
+	errg, ctx := errgroup.WithContext(ctx)
+	errg.Go(func() error {
+		return errors.Trace(m.run(ctx))
+	})
+
+	errg.Go(func() error {
+		return errors.Trace(m.pool.Run(ctx))
+	})
+
+	return errg.Wait()
+}
+
+func (m *MessageServer) run(ctx context.Context) error {
+	ticker := time.NewTicker(m.config.AckInterval)
+	defer ticker.Stop()
+
+	for {
+		failpoint.Inject("ServerInjectTaskDelay", func() {
+			log.Info("channel size", zap.Int("len", len(m.taskQueue)))
+		})
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-ticker.C:
```

Comment:
Could you use timer? Ticker may buffer a tick in its channel, as a result `m.tick()` may be invoked faster twice before next tick.

Reply:
The drawback to use ticker may cause unbalanced trigger, but the frequency can't exceed the interval. Such as trigger interval is 4s, it won't trigger more than 2 times during 8s, but internal between two ticks is not deterministic

Reply:
I think using a ticker is enough for our purpose here.